ocaml/libs/tracing/propagator.ml

```diff
@@ -0,0 +1,106 @@
+(*
+ * Copyright (c) Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module type S = sig
+  type carrier
+
+  val inject_into : Tracing.TraceContext.t -> carrier -> carrier
+
+  val extract_from : carrier -> Tracing.TraceContext.t
+end
+
+let ( let* ) = Option.bind
+
+let ( >> ) f g x = g (f x)
+
+let maybe f = function Some _ as o -> f o | _ -> Fun.id
+
+let[@tail_mod_cons] rec filter_append p xs ys =
+  match xs with
+  | [] ->
+      ys
+  | x :: xs when p x ->
+      x :: filter_append p xs ys
+  | _ :: xs ->
+      filter_append p xs ys
+
+module Http = struct
+  type carrier = Http.Request.t
+
+  open struct
+    let hdr_traceparent = "traceparent"
+
+    let hdr_baggage = "baggage"
+  end
+
+  let alloc_assoc k kvs =
+    List.filter_map
+      (fun (key, value) -> if key = k then Some value else None)
+      kvs
+    |> function
+    | [] ->
+        None
+    | xs ->
+        Some xs
+
+  let parse =
```

Comment:
The code in xapi tends to avoid point-free style, while the composition is nice to use inside anonymous functions, it's less so here where it can be replaced by adding the parameter and using `|>`